File: array/binary-search.py
import random
def binarySearchRecursive(arr, element, low, high):
    while(low<=high): 
        mid = low + (high - low) // 2

        if arr[mid] == element:
            return mid
        elif arr[mid] > element:
            return binarySearchRecursive(arr, element, low, mid-1)
        else :
            return binarySearchRecursive(arr, element, mid+1, high)
    return -1;

def binarySearchIterative(arr, element):
    low = 0
    high = len(arr) - 1
    while(low <= high):
        mid = low + (high - low) // 2;
        if arr[mid] == element:
            return mid
        elif arr[mid] < element:
            low = mid + 1
        else: 
            high = mid - 1
    return -1


def sort(arr):
    for i in range(len(arr)):
        for j in range(i+1,len(arr)):
            if arr[i] > arr[j]:
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;

    return arr


if __name__ == "__main__":
    test1Array = [2,4,1,0,3,6,5,9,7,8];
    searchingValue = random.choice(test1Array)
    sortedArray = sort(test1Array[:])
    print(sortedArray)
    print("The search value is ",searchingValue)
    print("Binary Search Iterative ->",binarySearchIterative(sortedArray,searchingValue))

    print("Binary Search Recursive ->", binarySearchRecursive(sortedArray, searchingValue, 0, len(sortedArray)-1))

File: array/binary-search.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int binarySearchRecursive(std::vector<int> arr, int element, int low, int high){
    // int mid = (low + high)/2; it may cause int overflow for large array sets
    int mid = low + (high - low)/2;
    while(low<=high){
        if(arr[mid]==element){
            return mid;
        } 
        else if(arr[mid]>element){
            return binarySearchRecursive(arr, element, low, mid-1);
        }
        else {
            return binarySearchRecursive(arr, element, mid+1, high);
        }
    }
    return -1;
}

int binarySearchIterative(const std::vector<int> arr, int element){
    int low = 0;
    int high = arr.size() - 1;
    while(low < high) {
        int mid = low + (high - low)/2;

        if(arr[mid] == element) return mid;
        else if (arr[mid] < element) low = mid + 1;
        else high = mid -1;
    }
    return -1;
}

void display(std::vector<int> arr){
    std::cout << "------------------------" << std::endl;
    for(int i=0; i<arr.size(); i++){
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl << "------------------------" << std::endl;
}

int main(){
    std::vector<int> nums = { 1, 5, 8, 9, 6, 7, 3, 4, 2, 0 };
    int first = 0;
    int last = nums.size()-1;
    int valueSeaching = 1;
    display(nums);
    std::sort(nums.begin(), nums.end());
    display(nums);
    std::cout << "Binary Search Recursive index value => " << binarySearchRecursive(nums, valueSeaching, first, last) << std::endl;
    std::cout << "Binary Search Iterative index value => " << binarySearchIterative(nums, valueSeaching) << std::endl;
    return 0;
}